### PR TITLE
ROS1Player publishing

### PR DESCRIPTION
**Pull request by jhurliman**
Advertise the publish capability from Ros1Player and implements setPublishers() and publish().

### CODE REVIEW

File: packages/ros1/src/TcpTypes.ts [hunk 1]
```diff
@@ -53,7 +53,7 @@ export interface TcpServerEvents {
 }
 
 export interface TcpServer {
-  address(): TcpAddress | undefined;
+  address(): Promise<TcpAddress | undefined>;
```
Comment (defunctzombie) [inline packages/ros1/src/TcpTypes.ts:56]: guessing this is intentional - why the change in this public interface? Are addresses for servers not known immediately? I guess not if the server hasn't started?

File: app/players/Ros1Player.ts [hunk 2]
```diff
@@ -342,10 +330,70 @@ export default class Ros1Player implements Player {
...
+    // Unadvertise any topics that were previously published and no longer appear in the list
+    for (const topic of this._rosNode.publications.keys()) {
+      if (!topics.has(topic)) {
+        this._rosNode.unadvertise(topic);
+      }
+    }
+
+    // Advertise new topics
+    for (const pub of publishers) {
+      if (!this._rosNode.publications.has(pub.topic)) {
```
Comment (defunctzombie) [inline app/players/Ros1Player.ts:370]: maybe flip this to early continue and less nesting?

File: app/players/Ros1Player.ts [hunk 3]
```diff
@@ -342,10 +330,70 @@ export default class Ros1Player implements Player {
...
+        const topic = pub.topic;
+        let msgdef: RosMsgDefinition[];
+        try {
+          msgdef = rosDatatypesToMessageDefinition(this._providerDatatypes, pub.datatype);
+        } catch (error) {
+          this._problems.push({
```
Comment (defunctzombie) [inline app/players/Ros1Player.ts:376]: When does this problem get cleared?

File: app/package.json [hunk 4]
```diff
@@ -68,7 +68,7 @@
     "regl-worldview": "github:foxglove/webviz#workspace=regl-worldview&commit=48979519bb7c067c7a0162255d1ef083fef49199",
     "rehype-raw": "5.1.0",
     "reselect": "4.0.0",
-    "rosbag": "github:foxglove/rosbag.js#fc61125fbc01ff55698c5ed48987b2d9e7599572",
+    "rosbag": "github:jhurliman/rosbag.js#c299e1ca95fc15b16e70465f10a28bcb1b350552",
```
Comment (jtbandes) [inline app/package.json:71]: 🤔 why do we need to switch from one fork to another?

File: app/players/Ros1Player.ts [hunk 5]
```diff
@@ -342,10 +330,70 @@ export default class Ros1Player implements Player {
...
+
   setPublishers(publishers: AdvertisePayload[]): void {
-    publishers = publishers.filter((p) => p.topic.length > 0);
-    if (publishers.length > 0) {
-      throw new Error("Publishing not supported");
+    if (!this._rosNode || this._closed) {
+      return;
+    }
+
+    // ROS1 only supports publishing topics that begin with "/"
```
Comment (jtbandes) [inline app/players/Ros1Player.ts:357]: what does this mean exactly? I know we've seen .bag files with non-prefixed topics in them. I think name resolution will make them resolve to the same thing (if a node is not namespaced) so why is publishing them not supported? This seems like a non-obvious limitation that would need to be documented somewhere for users at least, if it can't be removed

File: packages/ros1/src/nodejs/TcpServerNode.ts [hunk 6]
```diff
@@ -26,14 +26,14 @@ export class TcpServerNode extends EventEmitter<TcpServerEvents> implements TcpS
...
+  address(): Promise<TcpAddress | undefined> {
     const addr = this._server.address();
     if (addr == undefined || typeof addr === "string") {
       // Address will only be a string for an IPC (named pipe) server, which
       // should never happen in TcpServerNode
-      return undefined;
+      return Promise.resolve(undefined);
     }
-    return addr;
+    return Promise.resolve(addr);
```
Comment (jtbandes) [inline packages/ros1/src/nodejs/TcpServerNode.ts:36]: Prefer `async` functions rather than just returning Promise.resolve() everywhere

File: packages/ros1/src/TcpTypes.ts (hunk 1 above)
Comment (jhurliman) [reply packages/ros1/src/TcpTypes.ts:56]: It was inconsistent across the different implementations before, the code didn't compile (pass type checking). All of these methods have to return promises because the electron implementation requires async communication over postMessage
